modules/seqexec/server/src/main/scala/seqexec/server/gcal/GcalControllerEpics.scala

```diff
@@ -67,104 +41,93 @@ object GcalControllerEpics {
-    sys.diffuser.map(decode[String, Option[Diffuser]])
-
-  private def getFilter[F[_]: Monad](sys: GcalEpics[F]): F[Option[Filter]] =
-    sys.filter.map(decode[String, Option[Filter]])
-
-  private def getShutter[F[_]: Monad](sys: GcalEpics[F]): F[Option[Shutter]] =
-    sys.shutter.map(decode[String, Option[Shutter]])
-
-  private def getArLamp[F[_]: Sync](sys: GcalEpics[F]): F[Option[ArLampState]] =
-    sys.lampAr.map(v => ArLampState(decode(v)).some)
+  private def setArLampParams[F[_]: Async](sys: GcalEpics[F])(v: BinaryOnOff): F[Unit] =
+    sys.lampsCmd.setArLampName("Ar") *>
+      sys.lampsCmd.setArLampOn(v)
 
-  private def getCuArLamp[F[_]: Sync](sys: GcalEpics[F]): F[Option[CuArLampState]] =
-    sys.lampCuAr.map(v => CuArLampState(decode(v)).some)
+  private def setCuArLampParams[F[_]: Async](sys: GcalEpics[F])(v: BinaryOnOff): F[Unit] =
+    sys.lampsCmd.setCuArLampName("CuAr") *>
+      sys.lampsCmd.setCuArLampOn(v)
 
-  private def getQHLamp[F[_]: Sync](sys: GcalEpics[F]): F[Option[QHLampState]] =
-    sys.lampQH.map(v => QHLampState(decode(v)).some)
+  private def setThArLampParams[F[_]: Async](sys: GcalEpics[F])(v: BinaryOnOff): F[Unit] =
+    sys.lampsCmd.setThArLampName("ThAr") *>
+      sys.lampsCmd.setThArLampOn(v)
 
-  private def getThArLamp[F[_]: Sync](sys: GcalEpics[F]): F[Option[ThArLampState]] =
-    sys.lampThAr.map(v => ThArLampState(decode(v)).some)
+  private def setQHLampParams[F[_]: Async](sys: GcalEpics[F])(v: BinaryOnOff): F[Unit] =
+    sys.lampsCmd.setQHLampName("QH") *>
+      sys.lampsCmd.setQHLampOn(v)
 
-  private def getXeLamp[F[_]: Sync](sys: GcalEpics[F]): F[Option[XeLampState]] =
-    sys.lampXe.map(v => XeLampState(decode(v)).some)
+  private def setXeLampParams[F[_]: Async](sys: GcalEpics[F])(v: BinaryOnOff): F[Unit] =
+    sys.lampsCmd.setXeLampName("Xe") *>
+      sys.lampsCmd.setXeLampOn(v)
 
-  private def getIrLamp[F[_]: Sync](sys: GcalEpics[F]): F[Option[IrLampState]] =
-    sys.lampIr.map(v => IrLampState(decode(v)).some)
+  private def setIrLampParams[F[_]: Async](sys: GcalEpics[F])(v: BinaryOnOff): F[Unit] =
+    sys.lampsCmd.setIRLampName("IR") *>
+      sys.lampsCmd.setIRLampOn(v)
 
-  private def setArLampParams(v: BinaryOnOff): List[IO[Unit]] = List(
-    GcalEpics.instance.lampsCmd.setArLampName("Ar"),
-    GcalEpics.instance.lampsCmd.setArLampOn(v)
-  )
-
-  private def setCuArLampParams(v: BinaryOnOff): List[IO[Unit]] = List(
-    GcalEpics.instance.lampsCmd.setCuArLampName("CuAr"),
-    GcalEpics.instance.lampsCmd.setCuArLampOn(v)
-  )
-
-  private def setThArLampParams(v: BinaryOnOff): List[IO[Unit]] = List(
-    GcalEpics.instance.lampsCmd.setThArLampName("ThAr"),
-    GcalEpics.instance.lampsCmd.setThArLampOn(v)
-  )
+  def apply[F[_]: Async](epics: => GcalEpics[F]): GcalController[F] = new GcalController[F] {
+    override def applyConfig(config: GcalConfig): F[Unit] =
+      retrieveConfig(epics).flatMap(configure(epics, _, config))
```

Comment:
So we read the state and then do a smart configure?

Reply:
Yes, exactly.